kong/router/fields.lua

```diff
@@ -0,0 +1,322 @@
+      if not queries then
+        return true
+      end
+
+      for _, n in ipairs(v) do
+        local v = queries[n]
+        local f = "http.queries." .. n
+
+        -- the query parameter has only one value, like /?foo=bar
+        if type(v) == "string" then
+          local res, err = c:add_value(f, v)
+          if not res then
+            return nil, err
+          end
+
+        -- the query parameter has no value, like /?foo,
+        -- get_uri_arg will get a boolean `true`
+        -- we think it is equivalent to /?foo=
+        elseif type(v) == "boolean" then
+          local res, err = c:add_value(f, "")
+          if not res then
+            return nil, err
+          end
+
+        -- multiple values for a single query parameter, like /?foo=bar&foo=baz
+        elseif type(v) == "table" then
+          for _, v in ipairs(v) do
+            local res, err = c:add_value(f, v)
+            if not res then
+              return nil, err
+            end
+          end
+        end -- if type(v)
+      end   -- for ipairs(v)
+
+      return true
+    end,
+}
+
+
+function get_cache_key(fields, params)
+  for field, value in pairs(fields) do
+    local func = CACHE_KEY_FUNCS[field]
+
+    if func then
+      func(value, params, str_buf)
+    end
+  end
+
+  if not fields["http.host"] then -- preserve_host
+    CACHE_KEY_FUNCS["http.host"](true, params, str_buf)
+  end
+
+  if not fields["http.path"] then -- 05-proxy/02-router_spec.lua:1329
+    CACHE_KEY_FUNCS["http.path"](true, params, str_buf)
+  end
+
+  return str_buf:get()
+end
+
+
+else -- stream subsystem
+
+
+CACHE_KEY_FUNCS = {
+    ["net.src.ip"] =
+    function(v, params, buf)
+      buf:put(params.src_ip or ""):put("|")
+    end,
+
+    ["net.src.port"] =
+    function(v, params, buf)
+      buf:put(params.src_port or ""):put("|")
+    end,
+
+    ["net.dst.ip"] =
+    function(v, params, buf)
+      buf:put(params.dst_ip or ""):put("|")
+    end,
+
+    ["net.dst.port"] =
+    function(v, params, buf)
+      buf:put(params.dst_port or ""):put("|")
+    end,
+
+    ["tls.sni"] =
+    function(v, params, buf)
+      buf:put(params.sni or ""):put("|")
+    end,
+}
+
+
+MATCH_CTX_FUNCS = {
+    ["net.src.ip"] =
+    function(v, c, params)
+      return c:add_value("net.src.ip", params.src_ip)
+    end,
+
+    ["net.src.port"] =
+    function(v, c, params)
+      return c:add_value("net.src.port", params.src_port)
+    end,
+
+    ["net.dst.ip"] =
+    function(v, c, params)
+      return c:add_value("net.dst.ip", params.dst_ip)
+    end,
+
+    ["net.dst.port"] =
+    function(v, c, params)
+      return c:add_value("net.dst.port", params.dst_port)
+    end,
+
+    ["tls.sni"] =
+    function(v, c, params)
+      return c:add_value("tls.sni", params.sni)
+    end,
+
+    ["net.protocol"] =
+    function(v, c, params)
+      return c:add_value("net.protocol", params.scheme)
+    end,
+}
+
+
+function get_cache_key(fields, params)
+  for field, value in pairs(fields) do
```

Comment:
Again I don't think this is the correct way to generate a cache key. For the same input using pairs, there's no guarantee that the same order will apply, and it could be "192.168.1.1|80" or "80|192.168.1.1".

Reply:
I think that for a certain hash table each each iteration of `pair` should have the same order.
Is it correct?

Reply:
> For one Lua vm the order of iteration of pair() is decided, I think it is OK.

Where did you read this @chronolaw  🤔?

https://www.lua.org/manual/5.1/manual.html#pdf-next

> The order in which the indices are enumerated is not specified, even for numeric indices.

According to the lua 5.1 docs, the `pairs` doesn't guarantee the order of iteration.

Reply:
My perception is that depending on an undefined behavior is not a good idea (even if it is defined in a specified implementation), especially if it doesn't bring huge benefits.

Reply:
With `resty` the client tools, it runs in two different Lua VM, so the order may be different.

Reply:
https://github.com/Kong/kong/pull/12127#discussion_r1430822265

Reply:
Let's discuss there.